compiler/rustc_hir/src/hir.rs

```diff
@@ -3997,6 +3997,22 @@ mod size_asserts {
     // tidy-alphabetical-end
 }
 
+mod dropless {
+    use super::*;
+
+    macro_rules! static_assert_no_destructor {
+        ($ty:ty) => {
+            const _: () = {
+                if std::mem::needs_drop::<$ty>() {
+                    panic!(concat!(stringify!($ty), " has a destructor"));
+                }
+            };
+        };
+    }
+
+    static_assert_no_destructor!(Expr<'_>);
```

Comment:
I'm not sure whether this is a useful thing to have, but I used to to make sure that my code was correct and didn't want to immediately delete it

Reply:
maybe we should assert that all hir arena_types are dropless?

Reply:
this is kinda what #109590 does actually

Reply:
yeah, I'll just remove the assertion here, makes no sense to land, we should just land the other PR afterwards